### PR TITLE
Allow robots to be selected for tactics based on physical capability

### LINKED ISSUE #640: Allow robots to be selected for tactics based on physical capability
### Description of the task

Assign robots to tactics or explicitly avoid assigning them to tactics based on their physical capabilities.

### Outline the task in more detail

Robots inevitably break or sometimes have certain capabilities missing, such as not being able to dribble well or not being able to chip. We want to avoid assigning robots with these deficiencies to tactics that need those capabilities. ie. if a robot is not able to kick for whatever reason, we should avoid assigning it to a tactic that requires the ability to kick (like shooting).

Because our tactic assignment uses a "cost" system and the Hungarian optimization algorithm, the best approach will likely be to modify these costs before we apply the algorithm.

Possible solutions:
* Create an enum of flags that indicate robot capability (eg. kicking, chipping, dribbling, etc.) (is "moving" a valid capability?)
* Create some kind of dynamic parameters to indicate which robots don't have a certain capability
* Add a function to the abstract Tactic class that all tactics must implement. This function returns the capability flags that the tactic needs. Eg, the ShootTactic would return a "KickFlag" (and maybe a "ChipFlag").
* In the "assignRobotsToTactics" function in `stp.cpp`, get the capability flags a tactic wants along with the robot cost when creating the initial matrix (line 148). Using the dynamic parameters, check if the robot has the capability requested or not. If it does not, change the cost to be `1.0` (the highest cost) so we avoid assigning that robot to that tactic if possible.

### Acceptance criteria

- [ ] Do the above
- [ ] Test it

### PR DESCRIPTION
**Pull request by jordanzliu**
### Description

Created an enum of `RobotCapability` for dribbler, kicker, chipper, and a wrapper around `std::bitset` to represent a set of capabilities. 
Added `robotCapabilityRequirements` to `Tactic` with a default implementation of requiring nothing.
Changed tactic cost calculation in `stp.cpp` to set cost to `1.0` if the robot does not have at least the capabilities required by the tactic.

### Testing Done

made some unit tests
ran the modified kickoff play in grsim, works correctly (i.e. selects a farther robot if it's the only one with the required capabilities)

### Resolved Issues

Resolves #640 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/thunderbots/software/ai/util.h [hunk 1]
```diff
@@ -41,3 +43,16 @@
     {                                                                                    \
         return static_cast<CLASS>(~static_cast<UNDERLYING>(lhs));                        \
     }
+
+inline std::unordered_set<unsigned int> commaSeparatedListToSet(
+    const std::string& list_str)
+{
+    std::stringstream ss(list_str);
```
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/util.h:52]: Can we be more robust to strings not in csv format? Just logging and moving on is fine, but we don't want a stray character taking down AI (did this during testing)

File: src/thunderbots/software/util/ros_messages.cpp [hunk 2]
```diff
@@ -30,7 +32,11 @@ namespace Util
             return ball_msg;
         }
 
-        Robot createRobotFromROSMessage(const thunderbots_msgs::Robot& robot_msg)
+        Robot createRobotFromROSMessage(
+            const thunderbots_msgs::Robot& robot_msg,
+            const std::unordered_set<unsigned int>& broken_dribblers,
```
Comment (MathewMacDougall) [inline src/thunderbots/software/util/ros_messages.cpp]: The "broken" params are unused

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
This is great so far. Basically exactly what I was thinking of. Aside from the few other comments, it seems that all the robots have no capability by default? Based on the code, this doesn't seem to be the case but I printed in the `createTeamFromROSMessage` function and it said all robots were broken.

File: src/thunderbots/software/util/ros_messages.cpp (hunk 2 above)
Comment (jordanzliu) [reply src/thunderbots/software/util/ros_messages.cpp]: removed

File: src/thunderbots/software/ai/hl/stp/stp.cpp [hunk 3]
```diff
@@ -145,8 +145,19 @@ std::vector<std::shared_ptr<Tactic>> STP::assignRobotsToTactics(
...
         for (int col = 0; col < num_cols; col++)
         {
-            matrix(row, col) =
-                tactics.at(col)->calculateRobotCost(friendly_team_robots.at(row), world);
+            if (!friendly_team_robots.at(row).getRobotCapabilities().hasAllCapabilities(
+                    tactics.at(col)->robotCapabilityRequirements()))
+            {
+                // hardware requirements of tactic are not satisfied by the current robot
+                // set cost to 1.0f
+                matrix(row, col) = 1.0f;
```
[the remark this answers is not in the document]
Comment (jordanzliu) [reply src/thunderbots/software/ai/hl/stp/stp.cpp]: done

**Comment by jordanzliu** [comment]
fixed the thing where it said all robots were broken, turns out i had the signs the wrong way

**Review by garethellis0 — DISMISSED** [review]

**Review by henryymliu — DISMISSED** [review]
LGTM, nice!

**Review by MathewMacDougall — APPROVED** [review]
LGTM :+1:

**Pull request merged**